eip-0030.md

```diff
@@ -0,0 +1,329 @@
+    val noMoreTokens         = successor.tokens.size == 3
+  
+    val validStorageRent     = successor.value > minStorageRent
+
+    val reservesX0 = SELF.value
+    val reservesY0 = tokenY0._2
+    val reservesX1 = successor.value
+    val reservesY1 = tokenY1._2
+
+    val oraclePoolRateXY = oraclePoolBox.R4[Long].get 
+    val lpRateXY0 = reservesX0 / reservesY0  // we can assume that reservesY0 > 0 (since at least one token must exist) 
+    val lpRateXY1 = reservesX1 / reservesY1  // we can assume that reservesY1 > 0 (since at least one token must exist)
+    val isCrossing = (lpRateXY0 - oraclePoolRateXY) * (lpRateXY1 - oraclePoolRateXY) < 0 // if (and only if) oracle pool rate falls in between, then this will be negative
+     
+    val crossCounterIn = SELF.R5[Int].get
+    val crossCounterOut = successor.R5[Int].get
+    
+    val validCrossCounter = crossCounterOut == {if (isCrossing) crossCounterIn + 1 else crossCounterIn}
+     
+    val validRateForRedeemingLP = oraclePoolRateXY > lpRateXY0 * 9 / 10 // lpRate must be >= 0.9 oraclePoolRate // these parameters need to be tweaked
+    // Do we need above if we also have the tracking contract?
+     
+    val deltaSupplyLP  = supplyLP1 - supplyLP0
+    val deltaReservesX = reservesX1 - reservesX0
+    val deltaReservesY = reservesY1 - reservesY0
+    
+    // LP formulae below using UniSwap v2 (with initial token burning by bootstrapping with positive R4)
+    val validDepositing = {
+        val sharesUnlocked = min(
+            deltaReservesX.toBigInt * supplyLP0 / reservesX0,
+            deltaReservesY.toBigInt * supplyLP0 / reservesY0
+        )
+        deltaSupplyLP <= sharesUnlocked
+    }
+
+    val validRedemption = {
+        val _deltaSupplyLP = deltaSupplyLP.toBigInt
+        // note: _deltaSupplyLP, deltaReservesX and deltaReservesY are negative
+        deltaReservesX.toBigInt * supplyLP0 >= _deltaSupplyLP * reservesX0 && deltaReservesY.toBigInt * supplyLP0 >= _deltaSupplyLP * reservesY0
+    } && validRateForRedeemingLP
+
+    val validSwap =
+        if (deltaReservesX > 0)
+            reservesY0.toBigInt * deltaReservesX * feeNum >= -deltaReservesY * (reservesX0.toBigInt * feeDenom + deltaReservesX * feeNum)
+        else
+            reservesX0.toBigInt * deltaReservesY * feeNum >= -deltaReservesX * (reservesY0.toBigInt * feeDenom + deltaReservesY * feeNum)
+
+    val validAction =
+        if (deltaSupplyLP == 0)
+            validSwap
+        else
+            if (deltaReservesX > 0 && deltaReservesY > 0) validDepositing
+            else validRedemption
+
+    sigmaProp(
+        validSupplyLP1 &&
+        validSuccessorScript &&
+        validOraclePoolBox &&
+        preservedLpNFT &&
+        validLP &&
+        validY &&
+        noMoreTokens &&
+        validAction && 
+        validStorageRent && 
+        validCrossCounter
+    )
+}
+```
+## Tracking Contract
+
+```scala
+{
+  // Tracking box
+  //   R4: Crossing Counter of LP box
```

Comment:
Why do we need for the counter? I guess it would be enough to write down in register height when price condition violated (within certain margin), and then a special action can nullify the register if price is okay again. On receiver side (in the bank contract), we check that HEIGHT - trackingBox.R4 > threshold

Reply:
Yes, there can be other ways to achieve the goal stated in previous comment. So in the LP, we would need to store two values. One contains boolean when condition is violated and another height at which violated. 

We can consider this approach as well. That way we can skip "initiation" and "completion" of top-up and just have a single step.